src/merkle_tree.rs

```diff
@@ -125,13 +127,74 @@ impl MerkleTree {
         }
     }
 
-    #[cfg(test)]
     pub fn get_root(&self) -> Option<Hash> {
         match self.layers.first() {
             Some(root_layer) => root_layer.first().copied(),
             None => None,
         }
     }
+
+    fn get_leaves(&self) -> Option<&Vec<Hash>> {
+        self.layers.last()
+    }
+
+    pub fn generate_proof<T: std::convert::AsRef<[u8]>>(
+        &self,
+        elem: &T,
+    ) -> Option<(Vec<Hash>, usize)> {
+        let leaf_idx = self.get_leaf_idx(Sha3_256::digest(elem).into())?;
+        let mut proof = Vec::new();
+        let mut idx = leaf_idx;
+        let height = self.layers.len();
+        for layer in self.layers[1..height].iter().rev() {
+            if idx % 2 == 0 {
+                // If the right element does not exist we duplicate the left one and add it to the proof
+                let right_child = match layer.get(idx + 1) {
+                    Some(right) => right,
+                    None => &layer[idx],
+                };
+                proof.push(*right_child);
+            } else {
+                proof.push(layer[idx - 1]);
+            }
+            idx /= 2;
+        }
+
+        Some((proof, leaf_idx))
+    }
+
+    pub fn verify_proof<T: std::convert::AsRef<[u8]>>(
+        &self,
+        elem: &T,
+        proof: &Vec<Hash>,
+        mut idx: usize,
+    ) -> bool {
+        let mut hasher = Sha3_256::new();
+        hasher.update(elem);
+        let mut hash: Hash = hasher.finalize_reset().into();
+        for part in proof {
+            if idx % 2 == 0 {
+                hasher.update(hash);
+                hasher.update(part);
+                hash = hasher.finalize_reset().into();
+            } else {
+                hasher.update(part);
+                hasher.update(hash);
+                hash = hasher.finalize_reset().into();
+            }
```

Comment:
`hash = hasher.finalize_reset().into();` can just go after if, since you are calling it in both branches,

Reply:
Will change